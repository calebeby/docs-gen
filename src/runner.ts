import Project, { Node, CallExpression } from 'ts-simple-ast'
import { findRoutes, parseRoute, Route } from './docgen'
import { format } from './format'
import { printType } from './print-type'
import { parseType } from './parse-type'

interface RouteCollection {
  url: string
  routes: Route[]
}

const printFormatType = (n: Node) =>
  `
\`\`\`ts
${format(printType(parseType(n.getType(), n)))}\`\`\`
`

const printRoute = (route: Route) => {
  let text = `\n## \`${route.method.toUpperCase()}\`\n`
  if (route.comment) {
    text += '\n' + route.comment + '\n'
  }
  if (route.requestNode) {
    text += `
### Request
${printFormatType(route.requestNode)}\n`
  }
  if (route.responseNode) {
    text += `
### Response
${printFormatType(route.responseNode)}`
  }
  return text
}

const printRouteCollection = (routeCollection: RouteCollection) =>
  `# \`/${routeCollection.url}\`
${routeCollection.routes.map(printRoute).join('\n')}`

const printIndexEntry = (routeCollection: RouteCollection) => {
  const headerID = routeCollection.url.replace(/[{}/]/g, '').toLowerCase()
  return `- [\`/${routeCollection.url}\`](#${headerID})`
}

const mergeUrls = (routes: RouteCollection[] = [], currentRoute: Route) => {
  const matchingValue = routes.find(u => u.url === currentRoute.url)
  if (matchingValue === undefined) {
    routes.push({ url: currentRoute.url, routes: [currentRoute] })
  } else {
    matchingValue.routes.push(currentRoute)
  }
  return routes
}

export const printDocs = (input: CallExpression[]) => {
  const routeCollections = input
    .map(parseRoute)
    .filter((r): r is Route => r !== undefined)
    .reduce<RouteCollection[]>(mergeUrls, [])

  const index = routeCollections
    .map(printIndexEntry)
    .sort()
    .join('\n')

  const body = routeCollections
    .map(printRouteCollection)
    .sort()
    .join('\n')

  return '# Index\n\n' + index + '\n\n' + body
}

export const run = (input: string) => {
  const project = new Project()
  const file = project.createSourceFile('file.ts', input)
  return printDocs(findRoutes(file))
}
