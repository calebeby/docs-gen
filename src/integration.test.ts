import { run } from './runner'

test('type', () => {
  const input = `
interface BasicEventInfo {
  name: string // from TBA short name
  district?: string
  week?: number
  // UTC date
  startDate: string
  // UTC date
  endDate: string
  // Only if user is logged in
  starred?: boolean
  location: {
    lat: number
    lon: number
  }
}

export const getEvents = () => get<BasicEventInfo[]>('/events')

interface EventInfo extends BasicEventInfo {
  webcasts: {
    type: 'twitch' | 'youtube'
    url: string
  }
  location: {
    name: string
    lat: number
    lon: number
  }
}

export const getEventInfo = (eventKey: string) =>
  get<EventInfo>(\`/event/\${eventKey}/info\`)

interface Report {
  team: string
  stats: { [key: string]: Stat }
  notes: { [key: string]: string }
  eventKey: string
  matchKey: string
}

export const submitReport = (report: Report) =>
  put(
    \`events/\${eventKey}/matches/\${eventKey}_\${matchKey}/reports\`,
    report
  )
`
  expect(run(input)).toMatchSnapshot()
})
