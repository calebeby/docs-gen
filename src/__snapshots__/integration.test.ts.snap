// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`type 1`] = `
"# Index

- [\`/event/{eventKey}/info\`](#eventeventkeyinfo)
- [\`/events\`](#events)
- [\`/users/{userId}\`](#usersuserid)

# \`/event/{eventKey}/info\`

## \`GET\`

### Response

\`\`\`ts
type Data = {
  webcasts: {
    type: \\"twitch\\" | \\"youtube\\"
    url: string
  }
  location: {
    name: string
    lat: number
    lon: number
  }
  name: string
  district?: string
  week?: number
  // UTC date
  startDate: string
  // UTC date
  endDate: string
  // Only if user is logged in
  starred?: boolean
}
\`\`\`

# \`/events\`

## \`GET\`

comment
this is a comment on getEvents
another comment

### Response

\`\`\`ts
type Data = {
  name: string
  district?: string
  week?: number
  // UTC date
  startDate: string
  // UTC date
  endDate: string
  // Only if user is logged in
  starred?: boolean
  location: {
    lat: number
    lon: number
  }
}[]
\`\`\`

# \`/users/{userId}\`

## \`PATCH\`

### Request

\`\`\`ts
type Data = {
  username?: string
  firstName?: string
  lastName?: string
  password?: string
  roles?: {
    isAdmin: boolean
    isVerified: boolean
  }
}
\`\`\`


### Response

\`\`\`ts
type Data = null
\`\`\`
"
`;
